Sources/GRPC/Interceptor/ClientTransportFactory.swift

```diff
@@ -0,0 +1,275 @@
+        interceptedBy: interceptors,
+        onResponsePart
+      )
+      factory.configure(transport)
+      return transport
+    }
+  }
+}
+
+private struct HTTP2<Request, Response> {
+  /// The multiplexer providing an HTTP/2 stream for the call.
+  private var multiplexer: EventLoopFuture<HTTP2StreamMultiplexer>
+
+  /// The ":authority" pseudo-header.
+  private var authority: String
+
+  /// The ":scheme" pseudo-header.
+  private var scheme: String
+
+  /// An error delegate.
+  private var errorDelegate: ClientErrorDelegate?
+
+  /// A codec for serializing request messages and deserializing response parts.
+  private var codec: ChannelHandler
+
+  fileprivate init<Serializer: MessageSerializer, Deserializer: MessageDeserializer>(
+    multiplexer: EventLoopFuture<HTTP2StreamMultiplexer>,
+    scheme: String,
+    authority: String,
+    serializer: Serializer,
+    deserializer: Deserializer,
+    errorDelegate: ClientErrorDelegate?
+  ) where Serializer.Input == Request, Deserializer.Output == Response {
+    self.multiplexer = multiplexer
+    self.scheme = scheme
+    self.authority = authority
+    self.codec = GRPCClientCodecHandler(serializer: serializer, deserializer: deserializer)
+    self.errorDelegate = errorDelegate
+  }
+
+  fileprivate func makeTransport<Request, Response>(
+    to path: String,
+    for type: GRPCCallType,
+    withOptions options: CallOptions,
+    interceptedBy interceptors: [ClientInterceptor<Request, Response>],
+    _ onResponsePart: @escaping (ClientResponsePart<Response>) -> Void
+  ) -> ClientTransport<Request, Response> {
+    return ClientTransport(
+      details: self.makeCallDetails(type: type, path: path, options: options),
+      eventLoop: self.multiplexer.eventLoop,
+      interceptors: interceptors,
+      errorDelegate: self.errorDelegate,
+      onResponsePart
+    )
+  }
+
+  fileprivate func configure<Request, Response>(_ transport: ClientTransport<Request, Response>) {
+    transport.configure { _ in
+      self.multiplexer.flatMap { multiplexer in
+        let streamPromise = self.multiplexer.eventLoop.makePromise(of: Channel.self)
+
+        multiplexer.createStreamChannel(promise: streamPromise) { streamChannel in
+          streamChannel.pipeline.addHandlers([
+            _GRPCClientChannelHandler(
+              callType: transport.callDetails.type,
+              logger: transport.logger
+            ),
+            self.codec,
+            transport,
+          ])
+        }
+
+        // We don't need the stream, but we do need to know it was correctly configured.
+        return streamPromise.futureResult.map { _ in }
+      }
+    }
+  }
+
+  private func makeCallDetails(
+    type: GRPCCallType,
+    path: String,
+    options: CallOptions
+  ) -> CallDetails {
+    return .init(
+      type: type,
+      path: path,
+      authority: self.authority,
+      scheme: self.scheme,
+      options: options
+    )
+  }
+}
+
+private struct Fake<Request, Response> {
```

Comment:
Should we nest this and `HTTP2` inside extensions on `ClientTransportFactory`, just so their names are easier to identify in tracebacks and stuff?

Reply:
Hmm I had some weird compile errors when they were nested (with pre 5.3 compilers). I don't recall exactly what the error was but I'll make the names more obvious; `HTTP2ClientTransportFactory`.